### PR TITLE
GitHub Actions

### PR DESCRIPTION
**Pull request by ccjernigan**
The primary purpose of this is to set the stage for automated testing and deployment of snapshot releases to maven central, along with manual deployment of releases.  Most of the documentation on this process is in the commits.  The deployment is automatically disabled until the necessary credentials are added.

Addition of the detekt baseline allows detekt static analysis to also be run on each build.

### CODE REVIEW

**Comment by pacu** [comment]
adding @defuse and @mdr0id  from DevSecOps for review

File: .github/actions/setup/action.yml
Comment (defuse) [inline .github/actions/setup/action.yml:51]: We need to be careful that builds on the main branch don't use caches produced by potentially-malicious PRs, otherwise a malicious PR could insert malware into the cache that comes alive when a main branch build happens and steals the secrets.

I think [cache access restrictions](https://docs.github.com/en/actions/advanced-guides/caching-dependencies-to-speed-up-workflows#restrictions-for-accessing-a-cache) stop this from happening:

> Access restrictions provide cache isolation and security by creating a logical boundary between different branches. For example, a cache created for the branch feature-a (with the base main) would not be accessible to a pull request for the branch feature-b (with the base main).

This guarantees that PRs can't interfere with each other. It doesn't exactly say that main branch builds can't access PR branch caches but I think that's what it means.

Does that sound right? (It might be a good idea to mention this in the documentation.)

**Review by defuse — APPROVED** [review]
utACK modulo my question about caching.

File: .github/actions/setup/action.yml
Comment (ccjernigan) [reply .github/actions/setup/action.yml:51]: There's a hierarchy for cache restores.  My understanding is that branches can restore from the main branch but the main branch cannot restore from other branches, however the docs could probably more more explicit about this.  This what I read:  https://docs.github.com/en/actions/advanced-guides/caching-dependencies-to-speed-up-workflows#matching-a-cache-key

**Comment by mdr0id** [comment]
utACK; couple minor questions

1. What branch protection rules will be used with the above CI workflow and which status checks will be required? 

2. Will the test_robo_demo_app use different machines than the shared runners or will they all use shared runners for near term wallet needs?

File: .github/workflows/pull-request.yml
Comment (ccjernigan) [reply .github/workflows/pull-request.yml:253]: There are two things we could do to enhance this further:
 - Add a readme to the zip file so that someone downloading it would have a reminder visible to them
 - Our pull request checklist also covers this, although the checklist hasn't been added to this repo yet.  (Example is in the [Secant repo](https://github.com/zcash/secant-android-wallet/blob/main/.github/pull_request_template.md))

**Comment by ccjernigan** [comment]
> utACK; couple minor questions
> 
> 1. What branch protection rules will be used with the above CI workflow and which status checks will be required?

Suggestions: 
 - Require at least 1 review approval
 - Require the static analysis steps to succeed
 - Require the robo test to succeed (once we add the API key)
 - Require the test_android_modules to succeed (once we add the API key and once we fix the tests, as some of the tests are sort of broken right now)

Enhancements:
 - We can add a codeowners configuration, requiring additional approval of changes to the CI scripts

> 2. Will the test_robo_demo_app use different machines than the shared runners or will they all use shared runners for near term wallet needs?

There are two systems involved:
 - GitHub Actions runners for running the Gradle tasks; these should be pretty safe.  Each action step uses an ephemeral runner.  The cache is set up before this step, so this step shouldn't add anything back into the cache.  So it should be mostly independent from the other build steps
 - Firebase Test Lab provisions a temporary virtual device for us to use.  Its run in Google's cloud and should be very isolated.  We're relying on Google to manage setup and teardown in a secure way.  There's supposed to be no data carried over from one run to the next.  This should be significantly more secure than running the Android emulator locally.  (Also faster too, since the emulator run inside GitHub Actions won't support nested virtualization and therefore will be slow emulation of ARM instructions)

**Comment by ccjernigan** [comment]
Also some settings in GitHub Actions should probably be updated from their defaults:
 - Enable Allow select actions: Actions created by Github and Marketplace Actions by verified creators (allows the Google auth action)
 - Disable "run workflows from fork pull requests"
 - Change workflow permissions: Read repository contents permission.

It looks something like this

image

**Review by pacu — APPROVED** [review]

**Comment by mdr0id** [comment]
> > utACK; couple minor questions
> > 
> > 1. What branch protection rules will be used with the above CI workflow and which status checks will be required?
> 
> Suggestions:
> 
>     * Require at least 1 review approval
> 
>     * Require the static analysis steps to succeed
> 
>     * Require the robo test to succeed (once we add the API key)
> 
>     * Require the test_android_modules to succeed (once we add the API key and once we fix the tests, as some of the tests are sort of broken right now)
> 
> 
> Enhancements:
> 
>     * We can add a codeowners configuration, requiring additional approval of changes to the CI scripts
> 
> 
> > 2. Will the test_robo_demo_app use different machines than the shared runners or will they all use shared runners for near term wallet needs?
> 
> There are two systems involved:
> 
>     * GitHub Actions runners for running the Gradle tasks; these should be pretty safe.  Each action step uses an ephemeral runner.  The cache is set up before this step, so this step shouldn't add anything back into the cache.  So it should be mostly independent from the other build steps
> 
>     * Firebase Test Lab provisions a temporary virtual device for us to use.  Its run in Google's cloud and should be very isolated.  We're relying on Google to manage setup and teardown in a secure way.  There's supposed to be no data carried over from one run to the next.  This should be significantly more secure than running the Android emulator locally.  (Also faster too, since the emulator run inside GitHub Actions won't support nested virtualization and therefore will be slow emulation of ARM instructions)

Sounds reasonable. I have updated Github setting but have not changed the required status checks yet. Please let me know when you want to update these of if they block

**Pull request merged**
